ImageLounge/src/DkCore/DkImageStorage.cpp

```diff
@@ -327,41 +340,75 @@ QImage rotateImageFast(const QImage &img, double angle)
 
     if (angle == 90 || angle == 180 || angle == 270) {
         // Fast path
-        QImage imgIn = img;
-        if (imgIn.depth() > 32) {
-            // Ensure the data is 32-bit aligned
-            imgIn = imgIn.convertToFormat(QImage::Format_ARGB32);
-        }
 
 #ifdef WITH_OPENCV
-        QSize size = angle == 180 ? imgIn.size() : imgIn.size().transposed();
-        QImage imgOut = QImage(size, imgIn.format());
-        const cv::Mat matIn = cv::Mat(imgIn.height(), imgIn.width(), CV_8UC4, (uchar *)imgIn.constBits(), imgIn.bytesPerLine());
-        cv::Mat matOut = cv::Mat(imgOut.height(), imgOut.width(), CV_8UC4, imgOut.bits(), imgOut.bytesPerLine());
-
+        cv::RotateFlags rot{};
         if (angle == 180) {
-            cv::rotate(matIn, matOut, cv::ROTATE_180);
+            rot = cv::ROTATE_180;
         } else if (angle == 90) {
-            cv::rotate(matIn, matOut, cv::ROTATE_90_CLOCKWISE);
+            rot = cv::ROTATE_90_CLOCKWISE;
         } else {
-            cv::rotate(matIn, matOut, cv::ROTATE_90_COUNTERCLOCKWISE);
+            rot = cv::ROTATE_90_COUNTERCLOCKWISE;
         }
 
-        return imgOut;
+        // The pixel depth matters, but the color interpretation does not.
+        switch (img.depth()) {
+        case 1:
+            return rotateImageCVMat(img.convertToFormat(QImage::Format_Indexed8), rot, CV_8U);
+        case 8:
+            return rotateImageCVMat(img, rot, CV_8U);
+        case 16:
+            return rotateImageCVMat(img, rot, CV_16U);
+        case 32:
```

Comment:
I think this is only valid if the channels are 8-bit, there are a few Qt image formats have <8 bits per channel (e.g. RGB555). ~As written, the 16U case is only valid for 16-bit grayscale images.~ I think this might be OK as 16U ccovers  the entire pixel

You are also missing rgb/bgr-888  (24-bit)  which is fairly common unless that can't work for some reason.

Likewise, not all 32-bit format are 8bits per channel), there are a 10-bit variants for HDR.

Ok now I am thinking if you rotate the entire pixel instead of per-component this would cover most cases (except rgb-888). So CV_8UC4 should be CV_32U

Reply:
As transposing and flipping move pixel as a whole, the internal structure of the pixel does not matter. So using CV_8UC4 should have the same effect as CV_32U, because they are all 32-bit unit.

Here's the possible depths of QImage. Formats like RGB555 would have padding, see https://doc.qt.io/qt-6/qimage.html#Format-enum.
![image](https://github.com/user-attachments/assets/23ae987e-2885-4bf8-b09c-cacf8cae023f)

I missed the 24-bit one. Might take some time to do since there are no native 24-bit type...

Reply:
So long as opencv isn't rotating channels separately I guess you don't  need CV_32U, but that would be more transparent.  If the rotation angle was anything other than 90/180/270 then it wouldn't opencv have to subdivide and resample the pixels to get a nice result?